pkg/engine/engine.go

```diff
@@ -320,6 +330,40 @@ func (e *Engine) initialize(ctx context.Context, options ...Option) error {
 	return nil
 }
 
+func (e *Engine) initSourceManager(ctx context.Context) {
+	opts := []func(*sources.SourceManager){
+		sources.WithConcurrentSources(int(e.concurrency)),
+		sources.WithConcurrentUnits(int(e.concurrency)),
+		sources.WithSourceUnits(),
+		sources.WithBufferedOutput(defaultChannelBuffer),
+	}
+	if e.jobReportWriter != nil {
+		var mu sync.Mutex
+		unitHook := sources.NewUnitHook(ctx, sources.OnUnitHookFinish(func(metrics sources.UnitMetrics) {
+			metrics.Errors = common.ExportErrors(metrics.Errors...)
+			details, err := json.Marshal(map[string]any{
+				"version": 1,
+				"data":    metrics,
+			})
+			if err != nil {
+				ctx.Logger().Error(err, "error marshalling job details")
+				return
+			}
+			mu.Lock()
+			defer mu.Unlock()
+			if _, err := e.jobReportWriter.Write(details); err != nil {
+				ctx.Logger().Error(err, "error writing to file")
+				return
+			}
+			if _, err := e.jobReportWriter.Write([]byte{'\n'}); err != nil {
+				ctx.Logger().Error(err, "error writing to file")
+			}
```

Comment:
You could avoid this by creating `JobReportWriter.WriteLn` but I'm not sure how worth it that is

Reply:
Is `WriteLn` a standard interface or would that be custom? The nice thing about `io.Writer` is that a lot of structs already implement it (like `os.File`).

Reply:
oh, i missed that `jobReportWriter` is an interface. never mind. it kinda sucks that you have to check the error twice but i guess this is the best option.

Reply:
Oh maybe I could append a newline to `details` before writing it.